### PR TITLE
OCD-1096: fixed comparison for boolean value

### PR DESCRIPTION
**Pull request by dlucas1**
Still working on the unit test for this (it's not easy to unit test), but I want to get this out to Ashwini to finish up her testing.

### CODE REVIEW

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/certifiedProduct/CertifiedProductHandler2014.java [hunk 1]
```diff
@@ -699,7 +699,7 @@ private void parseTestTools(PendingCertifiedProductEntity product, PendingCertif
 				ttEntity.setTestToolVersion(testToolVersion);
 				TestToolDTO testTool = testToolDao.getByName(testToolName);
 				if(testTool != null) {
-					if(testTool.isRetired() && product.getIcs().equals("false")) {
+					if(testTool.isRetired() && product.getIcs() == false) {
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/certifiedProduct/CertifiedProductHandler2014.java]: In the parent class to this one I have a method "asBoolean". Would this have been of help to you? Users will often enter different boolean values 0, 1, true, false, tRuE, FALSE, etc and I tried to handle that variety

Comment (dlucas1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/certifiedProduct/CertifiedProductHandler2014.java]: The asBoolean() method gets called at CertifiedProductHandler2015.handle() line 78 in parseCertifiedProductDetails(). This is done before CertifiedProductHandler2015.parseCriteria().parseTestTools() at line 176. Thus, a check for asBoolean() in CertifiedProductHandler2014/2015.parseTestTools() would be redundant.

Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/upload/certifiedProduct/CertifiedProductHandler2014.java]: Got it! As discussed in slack, I think this logic actually needs to be moved into one of the validator classes. Leaving as changes requested for now.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/certifiedProduct/AmbulatoryModular2014Validator.java [hunk 2]
```diff
@@ -91,8 +91,9 @@ protected void validateDemographics(PendingCertifiedProductDTO product) {
 							product.getErrorMessages().add("There was no test tool found matching '" + testTool.getName() + "' for certification " + cert.getNumber() + ".");
 						} else {
 							TestToolDTO tt = ttDao.getById(testTool.getTestToolId());
-							if(tt != null && tt.isRetired()) {
-								product.getErrorMessages().add("Cannot use test tool '" + tt.getName() + "' since it is retired. Please choose a different test tool for " + cert.getNumber() + ".");
+							if(tt != null && tt.isRetired() && product.getIcs() == false) {
+								product.getErrorMessages().add("Test Tool '" + testTool.getName() + "' for Certification Result '" + cert.getNumber() 
+								+ "' has been retired for product with ICS=false.");
```
Comment (andlar) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/certifiedProduct/AmbulatoryModular2014Validator.java]: Can we change this line to something like `can not be used, as it is a retired tool, and this Certified Product does not carry ICS`? Anywhere we have `ICS=false` that phrase is too programmery for our end users

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/certifiedProduct/AmbulatoryModular2014Validator.java [hunk 3]
```diff
@@ -91,8 +93,25 @@ protected void validateDemographics(PendingCertifiedProductDTO product) {
 							product.getErrorMessages().add("There was no test tool found matching '" + testTool.getName() + "' for certification " + cert.getNumber() + ".");
 						} else {
 							TestToolDTO tt = ttDao.getById(testTool.getTestToolId());
-							if(tt != null && tt.isRetired()) {
-								product.getErrorMessages().add("Cannot use test tool '" + tt.getName() + "' since it is retired. Please choose a different test tool for " + cert.getNumber() + ".");
+							String[] idParts = product.getUniqueId().split("\\.");
+							if(idParts.length < 9) {
+								logger.error("CHPL ID must have 9 parts separated by '.'");
+							}
+							if(tt != null && tt.isRetired() && idParts[6].toString().equals("0")) {
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/certifiedProduct/AmbulatoryModular2014Validator.java]: Looks like if the id doesn't have 9 parts it logs an error but then continues on trying to access the 6th part. Should this be in an else or reorganized?

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/certifiedProduct/AmbulatoryModular2014Validator.java [hunk 4]
```diff
@@ -91,8 +93,25 @@ protected void validateDemographics(PendingCertifiedProductDTO product) {
...
-							if(tt != null && tt.isRetired()) {
-								product.getErrorMessages().add("Cannot use test tool '" + tt.getName() + "' since it is retired. Please choose a different test tool for " + cert.getNumber() + ".");
+							String[] idParts = product.getUniqueId().split("\\.");
+							if(idParts.length < 9) {
+								logger.error("CHPL ID must have 9 parts separated by '.'");
+							}
+							if(tt != null && tt.isRetired() && idParts[6].toString().equals("0")) {
+								Boolean containsInheritedCertificationStatus = false;
+								for (String error: product.getErrorMessages()) {
+									if(error.contains("Inherited Certification Status is true")){
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/validation/certifiedProduct/AmbulatoryModular2014Validator.java]: I don't think it's best practice to compare on this string if we decide to change the error message in the future.  A better way might be to set a protected field in the parent class where the error about ICS conflict is originally generated and just check that. Then you won't have to worry about parsing the ID into its separate parts either.

File: chpl/chpl-service/src/main/java/gov/healthit/chpl/web/controller/CertifiedProductController.java [hunk 5]
```diff
@@ -195,17 +195,17 @@
...
-		if(updateIcs == false && existingProductIcs == true && hasRetiredTestTool == true){
-			List<String> retiredTestToolNumbers = new ArrayList<String>();
-			for(CertificationCriterionEntity cce : retiredTestTools){
-				retiredTestToolNumbers.add(cce.getNumber());
-			}
-			
-			updateRequest.getErrorMessages().add("Cannot set ICS to false for a Certified Product with ICS=true and attested criteria that have a retired Test Tool. "
-					+ "The following are attested criteria that have a retired Test Tool: " + retiredTestToolNumbers.toString());
-		}
+//		Boolean updateIcs = updateRequest.getIcs();
```
Comment (kekey1) [inline chpl/chpl-service/src/main/java/gov/healthit/chpl/web/controller/CertifiedProductController.java]: delete :)

**Pull request merged**